### PR TITLE
Implement global & extensible interfaces

### PR DESCRIPTION
**Pull request by gpotter2**
A pretty major refactor based on a [very old PR](https://github.com/secdev/scapy/pull/2264), but pretty cool. This:

- unifies how interfaces work on all platforms using a `scapy/interfaces.py` file and introducing `conf.ifaces`
- makes the switch to libpcap work in all cases
- drops the last places where "dnet" was used to make it clear we stopped supporting it (env var, "pcapdnet" file)
- merges `get_if_list` on all platforms
- improves how interfaces are detected on OSX
- gives extra detail about the status of the interfaces using interfaces flags and lists all interfaces in a human friendly way
- allows multiple interfaces providers to cohabitate, in a [wireshark extcap](https://www.wireshark.org/docs/man-pages/extcap.html) style, and for each one to provide it's own sockets. This is very useful for USB sockets for instance, or if the user wants to add its own interfaces/providers to scapy.

It brings the "interfaces" objects to other platforms. Those are cool because they give you the content of the flags and avoid having to re-read the /etc/ files each time you need them.

### Screenshots

e.g.: Linux
image

e.g.: Usbpcap AND libpcap
image

This is fully backward compatible: the interfaces objects implement the same functions as strings: `conf.iface + "test" == "eth0test"` so that it doesn't break anything, and all functions use a `network_name` function to get the network name from the object or a string (instead of the current `if WINDOWS:` .... checks)

## Tested

- [x] Windows & Linux (manual + unit tests)
- [x] OS x (manual + unit tests)
- [x] *BSD (manual)

## How to test this

**Mainteners, please test the changes (there are unit tests, but I'd like you all to give some feedback on the features)**: try playing around with `conf.ifaces` and `conf.iface`, then with `conf.use_pcap` to understand what it does.

### CODE REVIEW

**Comment by gpotter2** [comment]
The PR is finished and seems stable enough. Now let's wait for 2.4.4 to be published.

**Comment by gpotter2** [comment]
@guedou @p-l- Feel free to review this ! You probably want to download it to play around

**Comment by p-l-** [comment]
Seems really nice! Do you think it would be possible to use `NetworkInterface` instances in routes?

**Comment by p-l-** [comment]
Also, maybe when assigning `conf.iface = "eth0"` Scapy could check if it can find such an interface and use the appropriate object instead of the string?

**Comment by gpotter2** [comment]
> Seems really nice! Do you think it would be possible to use NetworkInterface instances in routes?

I actually tried to do that at first but ran into some import loops (since `conf.route` is  also the place where interfaces historically got loaded) i just couldn't get rid of.

> Also, maybe when assigning conf.iface = "eth0" Scapy could check if it can find such an interface and use the appropriate object instead of the string?

That's a good idea, I'll try to do that.

**Comment by p-l-** [comment]
I like it! I'll need some time to test it if you don't need it merged soon.

**Comment by guedou** [comment]
I will have a look a the PR and try to play with it. Did you manage to test it successfully under FreeBSD, OpenBSD and NetBSD?

**Comment by guedou** [comment]
While testing the PR with macOS 10.14.6, I noticed that no IPv6 is displayed while displaying `conf.iface`. That is likely a bug.

**Comment by gpotter2** [comment]
Thanks. I also didn't test in on the BSDs

File: scapy/config.py
Comment (guedou) [inline scapy/config.py:555]: s/See conf.ifaces/See conf.ifaces output/

File: scapy/interfaces.py
Comment (guedou) [inline scapy/interfaces.py:32]: We should use `NotImplementedError` instead.

File: scapy/interfaces.py
Comment (guedou) [inline scapy/interfaces.py:61]: I think that we need to display all addresses, or find a good compromise. Currently, for IPv6, this only displays link-local addresses (fe80::/64) which is a bit confusing.

File: scapy/libs/winpcapy.py
Comment (guedou) [inline scapy/libs/winpcapy.py:70]: I don't understand this comment.

**Review by guedou — CHANGES_REQUESTED** [review]
Really cool PR!
We definitely need to test it on *BSD before merging.

Concerning unit tests, this is a good opportunity to move interfaces related test out of regression.uts.

File: scapy/arch/bpf/core.py
Comment (gpotter2) [reply scapy/arch/bpf/core.py:196]: Setting this to False would make no interface appear when running as non sudo.

File: scapy/interfaces.py
Comment (gpotter2) [reply scapy/interfaces.py:99]: For IPv6, this code probably needs improvement (there will probably be multiple IPv6 addresses. What should be considered the defaul tone?)

File: scapy/arch/bpf/core.py
Comment (gpotter2) [reply scapy/arch/bpf/core.py:61]: This would start getting too long. IMHO `network_name` isn't too bad

File: scapy/arch/linux.py
Comment (gpotter2) [reply scapy/arch/linux.py:104]: There is now a generic `get_if_list` function in `scapy.interfaces`

**Comment by gpotter2** [comment]
Should we just rename "bpf" to "bsd". I never understood this naming

File: scapy/interfaces.py
Comment (guedou) [reply scapy/interfaces.py:99]: Difficult question. It will depend on the routing table, and we will need to fallback to use the same information as the 'Src candidates' output of `conf.routes6`. I think that it will be better to try to display all of them, or have a crop mechanism such as done with `conf.routes6`.

File: scapy/arch/bpf/core.py
Comment (gpotter2) [reply scapy/arch/bpf/core.py:151]: We could add it back (though because this is BSD-specific I hope no one relied on it). What do you think

File: scapy/arch/bpf/core.py
Comment (guedou) [reply scapy/arch/bpf/core.py:63]: My propositions attempt to clarify `ifname = network_name(ifname)`. For example, I will prefer to have `ifname = resolve_interface_name(ifname)`.

File: scapy/arch/linux.py
Comment (guedou) [reply scapy/arch/linux.py:104]: Got it!

**Comment by guedou** [comment]
BPF is the name of the kernel interface used to manipulate raw interfaces (see https://www.freebsd.org/cgi/man.cgi?bpf(4)). IMHO. that's the correct way to name this directory.

File: scapy/arch/bpf/core.py
Comment (gpotter2) [reply scapy/arch/bpf/core.py:64]: We already had a libpcap specific `pcap_name` function which did the same purpose but for libpcap, which explains the naming. I see your point but we already have `resolve_iface` which converts [Interface, interface name, interface network name] -> Interface object (the opposite of `network_name`), and I find the names too similar.

File: scapy/arch/bpf/core.py
Comment (guedou) [reply scapy/arch/bpf/core.py:151]: Do you think that we could keep it and add a deprecation warning?

**Comment by gpotter2** [comment]
Alright that makes sense !

**Comment by guedou** [comment]
The display looks way better now. Thanks! I did not notice it during my first test, but on macOS inactive interfaces are also listed. I think that is not necessary as they Scapy won't be able to use them. Another point, the "WARNING: No IPv4 address found" messages are not necessary too.

**Comment by guedou** [comment]
On a fresh Ubuntu 18.04 installation, the following creating a dummy network interface with `sudo ip link add name scapy0 type dummy` gives the following output:
```
>>> conf.ifaces
Source  Index  Name    MAC                IPv4          IPv6                               
sys     3      scapy0  ce:d1:6c:1f:ac:32  0.0.0.0                                          
sys     2      eth0    00:1c:42:69:bc:31  10.211.55.11  fdb2:2c26:f4e4:0:21c:42ff:fe69:bc31
                                                        fe80::21c:42ff:fe69:bc31           
sys     1      lo      00:00:00:00:00:00  127.0.0.1     ::1 
```

I see several issues here:
- scapy0 is not up and should not be displayed
- the IPv4 should be empty

**Comment by guedou** [comment]
04b8d1eaef3b7d1228b3673b3a45059c8746a275 works fine on FreeBSD 12.1

**Comment by gpotter2** [comment]
I'm pretty sure `get_if_list()` used to display all interfaces, even the ones not running.

Should we remove interfaces that are not up from `get_if_list()` (everywhere in scapy: routes, conf.ifaces, ...), or only from the output of `conf.ifaces` display?

File: scapy/arch/libpcap.py
Comment (gpotter2) [reply scapy/arch/libpcap.py:341]: This part is very old. I can't remember. I remember adding exceptions gradually until I got fed up

File: scapy/arch/bpf/core.py
Comment (gpotter2) [reply scapy/arch/bpf/core.py:151]: I've added it to `scapy/interfaces.py`. Is that alright?

**Comment by guedou** [comment]
Let's keep the current `get_if_list()` behavior for now. I think that `conf.ifaces` should only list interfaces that Scapy could use.

By the way, do you think that `conf.ifaces` could be ordered by the interface index? `Index` could be the first column. An interesting use case could be `conf.iface = conf.ifaces[5]`.

File: scapy/arch/bpf/core.py
Comment (guedou) [reply scapy/arch/bpf/core.py:151]: Sounds good! Thanks.

File: scapy/arch/libpcap.py
Comment (guedou) [reply scapy/arch/libpcap.py:341]: Ok, Understood.

**Comment by gpotter2** [comment]
I've settled on a middle ground:

```
>>> conf.ifaces
Source  Index  Name  MAC                IPv4       IPv6
sys     2      eth0  Microsof:aa:aa:aa  10.0.0.5   XXXXX
sys     1      lo    00:00:00:00:00:00  127.0.0.1  ::1
```

```
>>> conf.ifaces.show(hidden=True)
Source  Index  Name    MAC                IPv4       IPv6
sys     4      scapy0  9a:4c:10:71:24:07
sys     2      eth0    Microsof:aa:aa:aa  10.0.0.5   XXXXX
sys     1      lo      00:00:00:00:00:00  127.0.0.1  ::1
```
I also changed the sorting

File: scapy/arch/bpf/core.py
Comment (gpotter2) [inline scapy/arch/bpf/core.py:140]: TODO extend message. e.g.

Permission denied: could not open /dev/bpf%i. Make sure you are running Scapy using sudo !

This should help us avoiding the thousands of "It doesn't work, it says /dev/bpf" issues

Comment (guedou) [reply scapy/arch/bpf/core.py:140]: Sounds good.

**Comment by guedou** [comment]
@p-l- what's your call?

**Review by p-l- — APPROVED** [review]

**Comment by guedou** [comment]
Can I merge this or do you want me to sqash the commits?

File: scapy/arch/bpf/core.py
Comment (gpotter2) [reply scapy/arch/bpf/core.py:64]: Oh we didn't resolve this. What's your opinion

**Comment by guedou** [comment]
I still think that the current function name needs to be changed to something more meaningful =)

**Pull request merged**
